lib/rules/curly.js

```diff
@@ -309,6 +331,115 @@ module.exports = {
                 hasUnsafeIf(statement) && isFollowedByElseKeyword(node);
         }
 
+        /**
+         * Missing curly brace reporting location examples:
+         *
+         * The following location should be reported
+         *
+         * if (foo) bar()
+         *        ^^^
+         *
+         * while(foo) bar();
+         *          ^^^
+         *
+         * do foo() while(bar)
+         * ^^^
+         *
+         * for(;;) bar();
+         *       ^^^
+         *
+         * for (var foo of bar) console.log(foo)
+         *                    ^^^
+         *
+         * for (var foo in bar) console.log(foo)
+         *                    ^^^
+         *
+         * if (foo) {bar()} else bar(1);
+         *                  ^^^^
+         *
+         * Unexpected curly brace reporting location examples:
+         *
+         * The following location should be reported
+         *
+         * if (true) foo(); else { baz(); }
+         *                       ^
+         *
+         * for (;;) { foo(); }
+         *          ^
+         *
+         * while (bar) { foo(); }
+         *             ^
+         *
+         * do{foo();} while(bar);
+         *   ^
+         *
+         * for (var foo of bar) {console.log(foo)}
+         *                      ^
+         *
+         * if (foo) bar() else {bar(1)}
+         *                ^^^^
+         *
+         * Get the location of missing or non missing curly brace.
+         * @param {ASTNode} node the node to check.
+         * @param {string} name name to check.
+         * @param {boolean} isMissing is curly brace missing.
+         * @returns {Object} Position object.
+         */
+        function getReportLoc(node, name, isMissing) {
+            if (name === "else") {
+
+                if (!isMissing) {
+                    const token = sourceCode.getTokenAfter(getElseKeyword(node));
+
+                    return token.loc;
+                }
+
+                return getElseKeyword(node).loc;
+            }
+
+            if (isMissing) {
+
+                if (name === "for" || name === "do") {
+                    const token = sourceCode.getTokenBefore(node.body);
+
+                    return {
+                        start: token.loc.end,
+                        end: node.body.loc.start
+                    };
+                }
```

Comment:
This could return a zero-length range if the previous token and the body are adjacent:

```js
/* eslint curly: ["error"]*/

for(;;)doSomething();
```

I'm not sure was the idea to underline the previous token as well, or just the range after the previous token?

Reply:
It will show the next token

![image](https://user-images.githubusercontent.com/26347874/88462264-bba22180-cec7-11ea-9f7a-df87d6a521e7.png)

Reply:
The goal of #12334 is to avoid zero-length ranges, so it would be nice to find a solution that entirely fixes this rule, if possible.

Reply:
how about from the last `;` to `)` ?

Reply:
If the idea is to include the first token before the body in underlining, we could take its `start` location.

Reply:
Updated location

![image](https://user-images.githubusercontent.com/26347874/88826677-a76f6480-d1e6-11ea-9cd0-e459074c4595.png)